bpo-45696: Deep-freeze

gvanrossum opened the pull request:
See https://github.com/faster-cpython/ideas/issues/84

This seems to be 10% faster than current main[1], so an improvement over Eric's freeze.

Downside: the data segment grows from 490 KB to 1.3 MB. (We could get ~300 KB back by not including the marshalled-frozen data.)

Another downside, for developers: whenever you touch a source file, `_bootstrap_python` gets rebuilt, and that triggers a rebuild of all the deep-frozen modules. (See [comment below](https://github.com/python/cpython/pull/29118#discussion_r741324650) for a possible fix.)

Haven't gotten it working on Windows yet. (The magic is disabled there for now.) I might do that in a separate PR.

___
[1] Measured on an x86 Mac, with lto-pgo, this shaves 1.5-2 msec off a total run time of around 14 msec. I measured by using the subprocess to run `python -c pass` 100 times.

<!-- issue-number: [bpo-45696](https://bugs.python.org/issue45696) -->
https://bugs.python.org/issue45696
<!-- /issue-number -->

ericsnowcurrently committed:
Add the frozen stdlib modules.

ericsnowcurrently committed:
Freeze the stdlib modules loaded during startup.

ericsnowcurrently committed:
Remember the stdlib dir during startup.

ericsnowcurrently committed:
Add placeholders for stdlib modules.

ericsnowcurrently committed:
Identify whether or not the executed is running installed.

ericsnowcurrently committed:
Group the import-related configs.

ericsnowcurrently committed:
Add the -X frozen_modules CLI option.

ericsnowcurrently committed:
Default to "off" if in development.

ericsnowcurrently committed:
Remember if built with --enable-optimizations.

ericsnowcurrently committed:
Default to "on" for PGO builds.

ericsnowcurrently committed:
Ignore frozen modules depending on "-X frozen_modules".

gvanrossum committed:
First part of deepfreeze

Not generating Makefile yet

ericsnowcurrently committed:
Add _Py_GetMainConfig().

ericsnowcurrently committed:
Let _Py_GetStdlibDir() fall back to a config.

ericsnowcurrently committed:
Add _PyConfig_InitImportConfig().

gvanrossum committed:
Make deepfreeze actually work

Still not generating Makefile

ericsnowcurrently committed:
Add a NEWS entry.

gvanrossum committed:
Generate all deepfreeze items in Makefile.pre.in

gvanrossum committed:
Don't use GET_CODE in the extern declarations

gvanrossum committed:
Merge branch 'pr-28107' into deepfreeze

gvanrossum committed:
Exclude encodings module for now

gvanrossum committed:
Regenerate Python/frozen.c

gvanrossum committed:
Make [deep]freeze process mess noisy

gvanrossum committed:
Regenerated MSVC project files

gvanrossum committed:
Merge branch 'main' into deepfreeze

gvanrossum committed:
Make deep-freezing work again

gvanrossum committed:
Make deep-freezing *really* work

gvanrossum committed:
Merge branch 'main' into deepfreeze

gvanrossum committed:
Fix test_ctypes and struct_frozen example in docs

gvanrossum committed:
Merge branch 'main' into deepfreeze

gvanrossum committed:
Undo some badly merged files

gvanrossum committed:
Add dependency on bootstrap for _bootstrap_python

gvanrossum committed:
Remove Programs/_bootstrap_python in 'make clean'

gvanrossum committed:
Move _bootstrap_python to toplevel dir, to get build working

gvanrossum committed:
Regenerate Makefile.pre.in

tiran commented on Makefile.pre.in:
```diff
@@ -739,6 +764,194 @@ regen-test-frozenmain: $(BUILDPYTHON)
+		$(srcdir)/Tools/scripts/deepfreeze.py \
+		$(srcdir)/Lib/importlib/_bootstrap.py \
+		-m importlib._bootstrap \
+		-o $(srcdir)/Python/deepfreeze/importlib._bootstrap.c
```
Makefiles are a complicated mess. There are some subtle issues in this code that breaks out-of-tree builds. The general rule is: any existing source file must be prefixed with ``$(srcdir)/``. Any generated file must not have the prefix, so they get placed in the build tree instead of the source tree.

You can also make use of make autovars here. The ``$@`` is the target name (e.g. ``Python/deepfreeze/importlib._bootstrap.c``) and ``$<`` is the left-most dependency (e.g. ``$(srcdir)/Lib/importlib/_bootstrap.py``).

```suggestion
Python/deepfreeze/importlib._bootstrap.c: $(srcdir)/Lib/importlib/_bootstrap.py $(srcdir)/Tools/scripts/deepfreeze.py $(BOOTSTRAP)
	@echo "Deepfreezing $@ from $<"
	@$(BOOTSTRAP) \
		$(srcdir)/Tools/scripts/deepfreeze.py \
		$< \
		-m importlib._bootstrap \
		-o $@
```

Can we teach ``deepfreeze.py`` to deduce the module name from the source name? It allow us to simplify the make rules even further. It might be possible to use a Makefile substitution rule here, too.

I'm dropping the code here for future reference:

```
DEEPFREEZE_PYS = \
	$(srcdir)/Lib/importlib/_bootstrap.py \
	$(srcdir)/Lib/importlib/_bootstrap_external.py \
	$(srcdir)/Lib/zipimport.py

# Lib/egg/spam.py -> egg/spam.c -> egg.spam.c -> Python/deepfreeze/egg.spam.c
DEEPFREEZE_CS = $(addprefix Python/deepfreeze/,$(subst /,.,$(patsubst $(srcdir)/Lib/%.py,%.c,$(DEEPFREEZE_PYS))))
```

tiran replied:
I found a better way to simplify your build rules, https://github.com/python/cpython/compare/main...tiran:deepfreeze-make

The updates rules use patterns (``%``), autovars (``$@`` and ``$<``), and target local variables (``target : KEY=VALUE``) to avoid repetition. I had to rename the target file name, though.

gvanrossum replied:
I'm generating this part of `Makefile.pre.in` anyway and I'm not too keen on renaming the target files (or on using patterns), so I think I'll pass on that solution, but I think I've fixed the use of `$(srcdir)` and am now using `$@` and `$<` in the generated rules. I can now focus on fixing the 4 remaining tests.

But the real problem with this PR is that it slows down the build (much more than the current freeze approach) because whenever you touch a file (even files unrelated to freezing, say dictobject.c), the dependencies cause `_bootstrap_python` to be rebuilt and then to be used to regenerate all the deepfrozen .c files, which then have to be recompiled. And that takes several seconds and causes a lot of commands to scroll by, where previously it would just be recompile dictobject.c and relink.

The problem is that there *are* different types of changes that require regenerating and recompiling the deepfrozen files, e.g. if the bytecode compiler changes or any of the affected .py modules is changed. But it is difficult to distinguish between the different types of dependencies.

tiran replied:
> The problem is that there _are_ different types of changes that require regenerating and recompiling the deepfrozen files, e.g. if the bytecode compiler changes or any of the affected .py modules is changed. But it is difficult to distinguish between the different types of dependencies.

Every C code change of a core file triggers a rebuild of the bootstrap interpreter, which then triggers a refreeze of every Python file. Make also has order only prerequisites that don't trigger a rebuild. Try this:

```
Python/deepfreeze/importlib._bootstrap.c: $(srcdir)/Lib/importlib/_bootstrap.py $(srcdir)/Tools/scripts/deepfreeze.py $(PYTHON_HEADERS) | $(BOOTSTRAP)
```

The rule only triggers a rebuild when a header files changes, but not when the bootstrap interpreter is updated. It still ensure that the bootstrap interpreter is built before the target is executed.

gvanrossum replied:
This seems to be a [GNU Make feature](https://www.gnu.org/software/make/manual/html_node/Prerequisite-Types.html). Is that a problem for portability? It seems to work on Mac, but IIRC there are some minor (but supported) platforms with less featureful Make implementations.

Also, the definition of `PYTHON_HEADERS` is later in the file than my usage, and then that dependency seems to have no effect on this rule -- I discovered that if I touch `Include/Python.h`, it rebuilds `_bootstrap_python` but doesn't redo the deep-freezing. (But if I touch e.g. `Lib/os.py`, that file does get deep-frozen, so the rest of the rule does work.)

I think I'll keep this idea in mind but I'll first focus on fixing the failing tests, so I can then move on to running the buildbots.

gvanrossum committed:
Use 'make' shortcuts in deep-freeze rules

gvanrossum committed:
Fix test_import, test_regrtest, test_trace

We need to set co_filename of (deep)frozen modules to
<frozen {modname}> to keep remove_importlib_frames() working.

gvanrossum committed:
Fix out-of-tree builds

This fixes test_tools (in particular test_freeze_simple_script).
To make debugging this a little easier, in _run_quiet(), when
the subprocess fails, dump stdout and stderr.

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

methane commented on Tools/scripts/deepfreeze.py:
```diff
@@ -0,0 +1,421 @@
+    global verbose
+    args = parser.parse_args()
+    verbose = args.verbose
+    with open(args.file) as f:
```
```suggestion
    with open(args.file, encoding="utf-8") as f:
```

methane commented on Tools/scripts/deepfreeze.py:
```diff
@@ -0,0 +1,421 @@
+    outdir = os.path.dirname(output)
+    if outdir:
+        os.makedirs(outdir, exist_ok=True)
+    with open(output, "w") as file:
```
```suggestion
    with open(output, "w", encoding="utf-8") as file:
```

ericsnowcurrently commented on Python/bootstrap_frozen.c:
```diff
@@ -0,0 +1,45 @@
+
+/* Frozen modules bootstrap */
```
This file is essentially identical to Python/frozen.c if we cut out the non-bootstrap parts, right?  So could we just use frozen.c with something like `#ifndef FROZEN_BOOTRAP` around the non-bootstrap parts?

Otherwise I recommend generating the parts of this file with Tools/scripts/freeze_modules.py, the same as with frozen.c.  If we don't then this will break when someone adds another module to the `BOOTSTRAP` list in freeze_modules.py.

ericsnowcurrently commented on Tools/scripts/deepfreeze.py:
```diff
@@ -0,0 +1,421 @@
+        self.write('#include "internal/pycore_code.h"')
+        self.write("")
+
+    @contextlib.contextmanager
+    def indent(self) -> None:
+        save_level = self.level
+        try:
+            self.level += 1
+            yield
+        finally:
+            self.level = save_level
+
+    def write(self, arg: str) -> None:
+        self.file.writelines(("    "*self.level, arg, "\n"))
+
+    @contextlib.contextmanager
+    def block(self, prefix: str, suffix: str = "") -> None:
+        self.write(prefix + " {")
+        with self.indent():
+            yield
+        self.write("}" + suffix)
```
This Printer class is handy. 🙂

ericsnowcurrently commented on Tools/scripts/deepfreeze.py:
```diff
@@ -0,0 +1,421 @@
+
+    @contextlib.contextmanager
+    def block(self, prefix: str, suffix: str = "") -> None:
+        self.write(prefix + " {")
```
FWIW, at the call sites I kept expecting a trailing space on "prefix" instead of it always being added here.

ericsnowcurrently commented on Tools/scripts/deepfreeze.py:
```diff
@@ -0,0 +1,421 @@
+    def field(self, obj: object, name: str) -> None:
+        self.write(f".{name} = {getattr(obj, name)},")
+
+    def generate_bytes(self, name: str, b: bytes) -> str:
+        self.write("static")
+        with self.indent():
+            with self.block("struct"):
+                self.write("PyObject_VAR_HEAD")
+                self.write("Py_hash_t ob_shash;")
+                self.write(f"char ob_sval[{len(b) + 1}];")
+        with self.block(f"{name} =", ";"):
+            self.object_var_head("PyBytes_Type", len(b))
+            self.write(".ob_shash = -1,")
+            self.write(f".ob_sval = {make_string_literal(b)},")
+        return f"& {name}.ob_base.ob_base"
```
Just to be sure I understood this function correctly, the output is something like this:

```c
static
    struct {
        PyObject_VAR_HEAD
        Py_hash_t ob_shash;
        char ob_sval[LEN];
    }
NAME = {
    ...
    ob_shash = -1,
    .ob_sval = ...,
};
```

Is that right?

gvanrossum replied:
Yes. We can't use a predefined struct type because `ob_sval` is inline and has a different length for each byte string. (Same for unicode strings.) I indented the struct to call out that it's not the end. If this was hand-written code I'd probably insist on pre-declaring the struct but given that this is generated code I like this way better.

ericsnowcurrently replied:
Yeah, it reads well.

ericsnowcurrently commented on Tools/scripts/deepfreeze.py:
```diff
@@ -0,0 +1,421 @@
+            # print(f"Cache hit {key!r:.40}: {self.cache[key]!r:.40}")
+            return self.cache[key]
+        self.misses += 1
+        match obj:
```
nice! 🙂

ericsnowcurrently commented on Tools/scripts/deepfreeze.py:
```diff
@@ -0,0 +1,421 @@
+        return ret
+
+    def generate(self, name: str, obj: object) -> str:
+        # Use repr() in the key to distinguish -0.0 from +0.0
+        key = (type(obj), obj, repr(obj))
```
clever

ericsnowcurrently commented on Tools/scripts/deepfreeze.py:
```diff
@@ -0,0 +1,421 @@
+                val = self.generate_complex(name, z)
+            case frozenset(fs):
+                val = self.generate_frozenset(name, fs)
+            case builtins.Ellipsis:
+                return "Py_Ellipsis"
+            case None:
+                return "Py_None"
```
Interesting.  I hadn't realized that marshal doesn't explicitly handle the one remaining singleton,`NotImplemented`.

gvanrossum replied:
That's presumably because it never occurs as a constant in a code object.

ericsnowcurrently reviewed (approved):
Mostly LGTM

I left a couple minor comments that you can deal with as you see fit.

gvanrossum committed:
Open files in UTF-8 mode

Co-authored-by: Inada Naoki <songofacandy@gmail.com>

tiran commented on Tools/scripts/deepfreeze.py:
```diff
@@ -407,6 +407,9 @@ def main() -> None:
     output = args.output or modname + ".c"
+    outdir = os.path.dirname(output)
+    if outdir:
+        os.makedirs(outdir, exist_ok=True)
```
That's one way to solve the missing directory in OOT builds. :)

The correct, designated way is a bit more obscure. The ``configure.ac`` file has a list of source directories that are populated in OOT builds. You have to add the directory to ``SRCDIRS`` block.

gvanrossum replied:
Okay, I'll bite. Why is the obscure way better?

gvanrossum committed:
Solve out-of-tree build differently

gvanrossum committed:
Use _bootstrap_python to create pybuilddir.txt and sysconfig

(Adapted from @tiran)

gvanrossum committed:
Remove Python/deepfreeze contents

gvanrossum committed:
Merge remote-tracking branch 'origin/main' into deepfreeze

gvanrossum committed:
Don't use $@ and $<, they don't work on FreeBSD

gvanrossum committed:
Fix out of tree builds by adding $(srcdir)
